### PR TITLE
Implement transforms for distribution parameters

### LINKED ISSUE #55: Register .transform methods for distribution parameters
PR https://github.com/pytorch/pytorch/pull/4902

Each distribution should know the support of its parameters and results (e.g. `(-inf,inf)`, `(0,inf)`, `(0,1)`). This issue is about registering transform pairs for parameters (unlike distribution outputs, parameters do not need a `.log_abs_det_jacobian()` method and so do not need a full `Bijector` class).

See [relevant section in the Design Doc](https://docs.google.com/document/d/1wnABg0cdyaVMr-Xqz_brHBwnPJuzTeSk3Oqko2HHSfE/edit#heading=h.unemmumtnzfk). See also #56

### PR DESCRIPTION
**Pull request by fritzo**
This PR registers pseudoinverse pairs (`.to_unconstrained()`, `.to_constrained()`) that are used to transform top-level distribution parameters during optimization. These functions may map between spaces of different dimensionality (e.g. for `simplex`) and therefore are not bijections.

This PR does not register `Bijector`s for transforming sample results for use in HMC. This more complex task will be addressed in #56 and requires an implementation of `Bijector` classes with `.log_abs_det_jacobian()` methods.

This PR does not transform discrete variables, since they are not amenable to stochastic optimization.

See [Design Doc](https://docs.google.com/document/d/1wnABg0cdyaVMr-Xqz_brHBwnPJuzTeSk3Oqko2HHSfE/edit#heading=h.unemmumtnzfk)
Fixes #55

### CODE REVIEW

File: torch/distributions/constraints.py [hunk 1]
```diff
@@ -1,5 +1,20 @@
...
+- ``constraints.positive``
+- ``constraints.real``
+- ``constraints.simplex``
+- ``constraints.unit_interval``
+"""
 
+import torch
 
 __all__ = [
     'Constraint',
```
Comment (dustinvtran) [inline torch/distributions/constraints.py]: This is a bike-shedding question, but `Constraints` also incorporates unconstrained spaces like `real`. Is there a name conflict with other parts of PyTorch's API if we rename `Constraint` to `Type`? (maybe tensor type?)

If I understand correctly, the idea is to formalize the category that a value or object belongs to. `Type` captures this, and also connects to Python type annotations if we ever plan to annotate manipulations like a Bijector or Transform's domain/range.

File: torch/distributions/constraints.py [hunk 2]
```diff
@@ -81,7 +105,7 @@ def check(self, value):
         return (value % 1 == 0) & (value >= 0)
 
 
-class _IntegerInterval(Constraint):
+class integer_interval(Constraint):
```
Comment (dustinvtran) [inline torch/distributions/constraints.py]: In the docstrings, can you elaborate on how the domain is supported over varying dimensions? E.g., with `integer_interval`, it's `[lower_bound, upper_bound]` where `*_bound` are arbitrary ndarrays. With `boolean`, it's restricted to scalar 0 and 1.

File: torch/distributions/transforms.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+            u (Variable or Tensor): An unconstrained value.
+
+        Returns:
+            (Variable or Tensor): A constrained value.
+        """
+        raise NotImplementedError
+
+
+@register_transform(constraints.real)
+class IdentityTransform(Transform):
```
Comment (dustinvtran) [inline torch/distributions/transforms.py]: What's the difference between, e.g., `IdentityTransform` and an `IdentityBijector`? In general, are there left-inverse-only transforms that don't have a log det Jacobian we can implement as a Bijector?

File: torch/distributions/constraints.py [hunk 4]
```diff
@@ -44,13 +59,22 @@ def check(self, x):
 
 
 def is_dependent(constraint):
```
Comment (dustinvtran) [inline torch/distributions/constraints.py]: What are applications in mind for `is_dependent` and `dependent_property`?

Comment (fritzo) [reply torch/distributions/constraints.py]: They're useful in generic programming as a kind of NotImplemented or NeedMoreInfo class. Here's an example:
```py
def initialize_params(random_variables):
    params = {}
    for var_name, dist in random_variables.items():
        assert isinstance(dist, Distribution)
        for param_name, constraint in dist.params.items():
            name = '{}.{}'.format(var_name, param_name)
            if is_dependent(constraint):                                   # <------ Example
                raise ValueError('cannot initialize constraints for dependent param {}'.format(name))
            params[name] = transform(constraint).to_constrained(torch.normal(1))
    return params
```

File: torch/distributions/transforms.py (hunk 3 above)
Comment (fritzo) [reply torch/distributions/transforms.py]: `Transform` is just a much lighter weight thing with fewer guarantees than `Bijector`. The main examples are manifolds that are more naturally represented as subspaces than as a single chart. E.g. von-Mises or spherical von-Mises variables can be represented as 2 and 3-vectors rather than unrolled into a local angle. Less contrived is the `simplex` constraint. We usually use `softmax` to optimize an `n+1`-dimensional parameter space and then project (not biject) down to an `n` dimensional constrained space.

Maybe I'm taking `bijector` too literally? Am I missing something?

File: torch/distributions/transforms.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+        unit = (x - c.lower_bound) / (c.upper_bound - c.lower_bound)
+        return torch.log(unit / (1 - unit))
+
+    def to_constrained(self, u):
+        c = self.constraint
+        unit = torch.sigmoid(u)
+        return c.lower_bound + unit * (c.upper_bound - c.lower_bound)
+
+
+@register_transform(constraints.simplex)
```
Comment (fritzo) [inline torch/distributions/transforms.py]: @dustinvtran This is the main motivating example.

Comment (fritzo) [reply torch/distributions/transforms.py]: BTW Thanks so much for reviewing this!

File: torch/distributions/constraints.py (hunk 4 above)
Comment (dustinvtran) [reply torch/distributions/constraints.py]: That makes sense. I guess it's necessary because `_Dependent` isn't part of the public namespace.

That example is interesting. Because it seems like you _can_ register a transform that's a function of the dependent type's parameters.

File: torch/distributions/transforms.py (hunk 3 above)
Comment (dustinvtran) [reply torch/distributions/transforms.py]: I think that design decision makes sense. I guess in many cases the transform could be refactored to simply call a Bijector's forward and inverse under the hood. But in other settings, it can't be written as a Bijector (if we define a Bijector as a class that requires implementing forward, inverse, and log det Jacobian).

File: torch/distributions/constraints.py (hunk 1 above)
Comment (dustinvtran) [reply torch/distributions/constraints.py]: After conversation with @pravnar at PPS 2018, I guess another possible term is "base measure". https://pps2018.soic.indiana.edu/files/2017/12/abstract.pdf

(I'm personally happy to defer to whatever, `Constraints` included—only bringing up other possibilities.)

Comment (fritzo) [reply torch/distributions/constraints.py]: Hmm base measure makes sense for `.support` but I think it assumes too much structure for non-Bayesian treatment of parameters. E.g. in maximum likelihood there is no measure.

I kindof prefer `Constraint` rather than `Type` out of fear that these objects might attract additional type-like information (bit precision, cuda device, dimension, reparameterization type, is_exponential_family, ...). `Constraint` really emphasizes the non-goals :smile:

File: torch/distributions/transforms.py (hunk 3 above)
Comment (fritzo) [reply torch/distributions/transforms.py]: That's a great point: this could be a thin interface with only one or two non-`Bijector` pairs.

File: torch/distributions/constraints.py (hunk 1 above)
Comment (fritzo) [reply torch/distributions/constraints.py]: Oh @pravnar's approach might be useful for automatically inferring infinite KL's from base measure, as @vishwakftw and I have been discussing in #77

Comment (dustinvtran) [reply torch/distributions/constraints.py]: +1 to keeping `Constraint` and using it for infinite KLs!

**Comment by fritzo** [comment]
Closing in favor of #108
